### PR TITLE
Negative caching

### PR DESCRIPTION
**Pull request by khibino**
- [x] extend cache to save empty data.
- [x] extend lookup interface for empty data with SOA
- [x] caching empty NS in iterative steps
- [x] lookup empty NS in iterative steps
- [x] caching empty results
- [x] lookup empty for results

### CODE REVIEW

File: src/DNSC/Iterative.hs [hunk 1]
```diff
@@ -574,6 +576,47 @@ cacheSection rs rank = cacheRRSet
       insertRRSet <- asks insert_
       liftIO $ mapM_ ($ insertRRSet) rrss
 
+cacheEmptySection :: Domain -> Domain -> TYPE
```
Comment (kazu-yamamoto) [inline src/DNSC/Iterative.hs:646]: I would like to have a manual (comment) for this function.

File: src/DNSC/Iterative.hs [hunk 2]
```diff
@@ -575,6 +575,20 @@ lookupCache dom typ = do
                                         maybe "miss" (\ (_, rank) -> "hit: " ++ show rank) result]
   return result
 
+-- when cache has EMPTY result, lookup SOA data for domain delegation from
```
Comment (kazu-yamamoto) [inline src/DNSC/Iterative.hs:597]: Add a document with `--| `.

File: src/DNSC/Iterative.hs (hunk 1 above)
Comment (khibino) [reply src/DNSC/Iterative.hs:649]: https://github.com/khibino/dns-resolver/pull/26/commits/ea37a076991bb5364f9980905863c0da8aa8991a

File: src/DNSC/Iterative.hs (hunk 2 above)
Comment (khibino) [reply src/DNSC/Iterative.hs]: https://github.com/khibino/dns-resolver/pull/26/commits/77c20c6644528e0e43fba44ecdae34381fc47baf

**Review by kazu-yamamoto — APPROVED** [review]
LGTM

**Pull request merged**